### PR TITLE
Daprd: always use mTLS with control plane services when in kubernetes mode

### PR DESCRIPTION
**Pull request by JoshVanL**
Closes https://github.com/dapr/dapr/issues/6949

When running in Kubernetes mode, the control plane services (correctly) always network with [m]TLS, even when `mtls` has been
disabled in configuration.

Currently, daprd will (wrongly) attempt to plain text connect to control plane services when `mtls` has been disabled in configuration. This PR updates the `security` package so that it will always request it's identity certificate when running in Kubernetes mode, regardless of of the `mtls` config option. It has also been updated to that if the identity cert is available (`mtls=true` or `kubernetes=true`) mTLS Dial (which is used by daprd to connect to control plane services) is always mTLS.  

`mtls=false` will continue to allow Daprds to use insecure connections between each other.

### CODE REVIEW

**Review by ItalyPaleAle — DISMISSED** [review]
Sorry but I don't think this is a solution to the problem. if mTLS is disabled, it means users want it to be disabled everywhere, not just between sidecars. Having mTLS enabled comes with some additional complexities that may not be desirable, such as the need to rotate root certs...

I agree (m)TLS is much better in general, but we should honor what the user wants in this case.

**Comment by ItalyPaleAle** [comment]
> @ItalyPaleAle As I understand, it has always been the case that disabling `mTLS`, disables `mTLS` between daprd sidecards. `daprd => control plane` and `control plane => control plane` always enforces mTLS regardless (in Kubernetes).

I think you may be right. In 1.11 the only control plane service that had a `tls-enabled` CLI flag was Placement, but in the Helm chart it was always enabled. Ok, i'll dismiss my review

File: tests/integration/suite/daprd/mtls/kubernetes/operator.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	operatorv1pb "github.com/dapr/dapr/pkg/proto/operator/v1"
+	"github.com/dapr/dapr/pkg/security"
+	procgrpc "github.com/dapr/dapr/tests/integration/framework/process/grpc"
+)
+
+// operator is a mocked implementation of the operator service.
+type operator struct{}
```
Comment (JoshVanL) [inline tests/integration/suite/daprd/mtls/kubernetes/operator.go:32]: This stubs the operator sever with always enforced mTLS

**Comment by JoshVanL** [comment]
The PR with only the integrations tests has failed CI because of the following test. This is expected as this is the issue this PR is trying to address.

```
done
2023-09-21T13:44:36.8442449Z     --- FAIL: Test_Integration/daprd/mtls/kubernetes/disable (180.01s)
```

As you can see, this test is now passing with the implementation changes on this PR.

```
2023-09-21T13:42:52.5301368Z         --- PASS: Test_Integration/daprd/mtls/kubernetes/disable/run (2.31s)
```

File: pkg/security/x509source.go [hunk 2]
```diff
@@ -156,7 +157,7 @@ func newX509Source(ctx context.Context, clock clock.Clock, cptd spiffeid.TrustDo
 		appID:          opts.AppID,
 		appNamespace:   ns,
 		trustDomain:    trustDomain,
-		kubernetesMode: os.Getenv("KUBERNETES_SERVICE_HOST") != "",
+		kubernetesMode: opts.Mode == modes.KubernetesMode,
```
Comment (artursouza) [inline pkg/security/x509source.go:160]: Is this a breaking change? Shouldn't we use both flags to enable this (via an OR)?

Comment (JoshVanL) [reply pkg/security/x509source.go:160]: As discussed in the call yesterday, we agreed that the `KUBERNETES_SERVICE_HOST` environment variable is fallible so instead we are determining the mode via a passed option. For Daprd, this option is coming from the existing `--mode` CLI flag. For placement, this is coming from a new CLI flag `--mode`.

it's not a breaking change.

File: pkg/security/security.go [hunk 3]
```diff
@@ -135,8 +141,11 @@ func New(ctx context.Context, opts Options) (Provider, error) {
 		return nil, fmt.Errorf("invalid control plane trust domain: %w", err)
 	}
 
+	// Always request certificates from Sentry if mTLS is enabled or running in
+	// Kubernetes. In Kubernetes, Daprd always communicates mTLS with the control
+	// plane.
 	var source *x509source
-	if opts.MTLSEnabled {
+	if opts.MTLSEnabled || opts.Mode == modes.KubernetesMode {
```
Comment (yaron2) [inline pkg/security/security.go:148]: Thanks for making that change, this is much better.

**Review by yaron2 — APPROVED** [review]

**Pull request merged**